src/coreclr/src/jit/rangecheck.cpp

```diff
@@ -1117,8 +1164,8 @@ Range RangeCheck::ComputeRange(BasicBlock* block, GenTree* expr, bool monIncreas
         range = ComputeRangeForLocalDef(block, expr->AsLclVarCommon(), monIncreasing DEBUGARG(indent + 1));
         MergeAssertion(block, expr, &range DEBUGARG(indent + 1));
     }
-    // If add, then compute the range for the operands and add them.
-    else if (expr->OperGet() == GT_ADD)
+    // compute the range for binary operation
+    else if (expr->OperIs(GT_ADD, GT_AND, GT_RSH, GT_LSH, GT_UMOD))
```

Comment:
Seems like `GT_LSH` could overflow ... (?)

Reply:
@AndyAyersMS from the [spec (12.10)](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-334.pdf#page=201&zoom=100,73,513)

> Shift operations never cause overflows and produce the same results in checked and unchecked contexts.

Reply:
The spec is talking about overflow exceptions. The shifted value, when interpreted as a signed int, can "overflow" from positive to negative (and so no longer be monotonically increasing), which is what we care about here:
```
  int i = 1;
  while (i >= 0) i = i << 1;
  Console.WriteLine($"{i}");

  ;; prints -2147483648
```

Reply:
@AndyAyersMS the code calculates `<<` for two `int32` constants and if the result is negative - returns `Limit::keUnknown` https://github.com/dotnet/runtime/blob/40874004a1f2e423527220f26c31cf6e6395830e/src/coreclr/src/jit/rangecheck.cpp#L811-L814 (the test should contain cases for that)

Reply:
Hmm, maybe it's more of an issue with how range checking is structured. I would expect overflow checking to be integrated into the main range computation rather than being done after the fact.

So the contract for `ComputeDoesOverflow` is a bit odd; basically we're relying on the fact that we're careful elsewhere.

Reply:
@AndyAyersMS I can drop `GT_LSH` from there to keep things clean and logical, I wanted to handle `(x & c1) >> c2` which is used in some places and decided that I can also add `GT_LSH` for free.

Reply:
Maybe just add a comment to line 1077 explaining that this depends on the analysis done in `ComputeRangeForBinOp`

Reply:
Done.